homeassistant/components/bluetooth/manager.py

```diff
@@ -0,0 +1,393 @@
+"""The bluetooth integration."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable
+from datetime import datetime, timedelta
+import logging
+import time
+from typing import TYPE_CHECKING
+
+import async_timeout
+from bleak import BleakError
+from dbus_next import InvalidMessageError
+
+from homeassistant import config_entries
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import (
+    CALLBACK_TYPE,
+    Event,
+    HomeAssistant,
+    callback as hass_callback,
+)
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import discovery_flow
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.util.package import is_docker_env
+
+from . import models
+from .const import (
+    DEFAULT_ADAPTERS,
+    SCANNER_WATCHDOG_INTERVAL,
+    SCANNER_WATCHDOG_TIMEOUT,
+    SOURCE_LOCAL,
+    START_TIMEOUT,
+    UNAVAILABLE_TRACK_SECONDS,
+)
+from .match import (
+    ADDRESS,
+    BluetoothCallbackMatcher,
+    IntegrationMatcher,
+    ble_device_matches,
+)
+from .models import (
+    BluetoothCallback,
+    BluetoothChange,
+    BluetoothScanningMode,
+    BluetoothServiceInfoBleak,
+    HaBleakScanner,
+    HaBleakScannerWrapper,
+)
+from .usage import install_multiple_bleak_catcher, uninstall_multiple_bleak_catcher
+
+if TYPE_CHECKING:
+    from bleak.backends.device import BLEDevice
+    from bleak.backends.scanner import AdvertisementData
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+MONOTONIC_TIME = time.monotonic
+
+
+SCANNING_MODE_TO_BLEAK = {
+    BluetoothScanningMode.ACTIVE: "active",
+    BluetoothScanningMode.PASSIVE: "passive",
+}
+
+
+class BluetoothManager:
```

Comment:
Next PR will break this up some more so we have the overall manager and then a class per adapter/remote